music/src/core/data.ts

```diff
@@ -493,3 +501,242 @@ export class TrioConverter extends DataConverter {
+          0);
+
+      // Add tokens for each performance event.
+      track.events.forEach((event, index) => {
+        switch (event.type) {
+          case 'note-on':
+            tokens.set(event.pitch - constants.MIN_MIDI_PITCH, index + 1);
+            break;
+          case 'note-off':
+            tokens.set(
+                this.numPitches + event.pitch - constants.MIN_MIDI_PITCH,
+                index + 1);
+            break;
+          case 'time-shift':
+            tokens.set(2 * this.numPitches + event.steps - 1, index + 1);
+            break;
+          case 'velocity-change':
+            tokens.set(
+                2 * this.numPitches + this.totalSteps + event.velocityBin - 1,
+                index + 1);
+            break;
+          default:
+            throw new Error(`Unrecognized performance event: ${event}`);
+        }
+      });
+
+      // Add a single end token.
+      tokens.set(this.depth - 1, track.events.length + 1);
+    } else {
+      // Track doesn't exist, just use an end token.
+      tokens = tf.buffer([1], 'int32', new Int32Array([this.depth - 1]));
+    }
+
+    // Now do one-hot encoding and pad with zeros up to the maximum number of
+    // events.
+    const oh = tf.oneHot(tokens.toTensor() as tf.Tensor1D, this.depth);
+    return oh.pad([[0, maxEventsPerTrack - oh.shape[0]], [0, 0]]);
+  }
+
+  toTensor(noteSequence: INoteSequence) {
+    sequences.assertIsRelativeQuantizedSequence(noteSequence);
+
+    if (noteSequence.quantizationInfo.stepsPerQuarter != this.stepsPerQuarter) {
+      throw new Error(`Steps per quarter note mismatch: ${
+          noteSequence.quantizationInfo.stepsPerQuarter} != ${
+          this.stepsPerQuarter}`);
+    }
+
+    const instruments =
+        new Set(noteSequence.notes.map(note => note.instrument));
+    const tracks =
+        Array.from(instruments)
+            .map(
+                instrument => performance.Performance.fromNoteSequence(
+                    noteSequence, this.totalSteps, this.numVelocityBins,
+                    instrument));
+    const sortedTracks = tracks.sort(
+        (a, b) => b.isDrum ? -1 : (a.isDrum ? 1 : a.program - b.program));
+
+    // Drop tracks until we have the maximum number of instruments.
+    while (sortedTracks.length > this.numSegments) {
+      sortedTracks.pop();
+    }
+
+    // Make sure all tracks are the proper length (in time).
+    sortedTracks.forEach((track) => track.setNumSteps(this.totalSteps));
+
+    // Pad with "undefined" tracks to reach the desired number of instruments.
+    while (sortedTracks.length < this.numSegments) {
+      sortedTracks.push(undefined);
+    }
+
+    // Convert tracks to tensors then concatenate.
+    return tf.tidy(
+        () => tf.concat(
+            sortedTracks.map((track) => this.trackToTensor(track)), 0));
+  }
+
+  private tokensToTrack(tokens: Int32Array) {
+    // Trim to end token.
+    const idx = tokens.indexOf(this.depth - 1);
```

Comment:
I think it would be useful to store the end token as a property for readability.

Reply:
Done.